### ISSUE TITLE
This program using recurse hogs CPU and memory when compiled to AGC

### ISSUE THREAD

**Issue by MareinK** [alantech/alan#317]
When I `alan compile` to AGC and then `alan run` the following program, the `alan` process starts taking a huge amount of CPU and memory, starting to crash my computer before I terminate the process. The problem is caused by the executions on the last `print` line. The program works correctly when compiled for JS and ran with `node`. The problem is not resolved by changing `triangularRec` to from inline to 'bracketed' function syntax.

```
import @std/app
from @std/seq import seq, Self, recurse

fn triangular(x: int) : int = div(x ** 2 + x, 2)

fn triangularRec(x: int) : int = seq(x).recurse(fn (self: Self, x: int) : Result<int> {
    if x == 0 {
        return ok(x)
    }
    else {
        return ok(x + (self.recurse(x - 1) || 0))
    }
}, x) || 0

on app.start {
    app.print(triangularRec(1)) // 1
    app.print(triangularRec(2)) // 3
    app.print(triangularRec(3)) // 6

    const xs = [1, 2, 3]
    app.print(xs.map(triangular).map(toString).join(" ")) // 1 3 6
    app.print(xs.map(triangularRec).map(toString).join(" ")) // 1 3 6 with js, problem with agc

    emit app.exit 0
}
```

**Comment by dfellis** [comment]
There were actually two bugs in here that caused problems. After I fixed the main bug that was causing the infinite looping (not updating the recursion termination check correctly in parallel situations) I was getting weird results. The output from the AVM was `1 4 9` instead of `1 3 6`.

It was adding the original number `n` times instead of decrementing it as expected.

While trying to debug that, I realized there was another bug with setting up the arguments for the recursive function itself that could cause the wrong value to be pulled, and it was pulling an array value and placing it where the code expects a fixed numeric value.

After I fixed that, I realized that that sort of bug should've prevented the code from working at all, and then I finally realized there's an aliasing issue in the `ammtoaga` layer of the compiler. `triangularRec` aliases the `x` variable, but the scope evaluation in the aliasing was pulling the wrong value. Replacing the inner `x` with `y` fixes the runtime.

I'm now digging into the compiler to try and fix that aliasing bug, but I'm not sure how long this will take, so I'm putting up my notes now in case I need to continue on Monday (my weekend is packed so I won't be able to touch anything then).
